src/core/gatherKeys.js

```diff
@@ -0,0 +1,24 @@
+export const gatherKeys = template => {
+  const variablePositions = [];
+  const variablePattern = /{{\s*(?<key>\w+)\s*}}/g;
+  let match;
+  let offset = 0;
+
+  // eslint-disable-next-line no-cond-assign
+  while ((match = variablePattern.exec(template))) {
```

Comment:
I don't personally find this very readable. Could you extract this into a more explicit loop/condition?